File: app/src/main/java/de/devmil/paperlaunch/view/LaunchLaneView.java
/*
 * Copyright 2015 Devmil Solutions
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package de.devmil.paperlaunch.view;

import android.animation.Animator;
import android.animation.ObjectAnimator;
import android.content.Context;
import android.graphics.Color;
import android.graphics.Rect;
import android.graphics.drawable.Drawable;
import android.support.v7.graphics.Palette;
import android.util.AttributeSet;
import android.util.TypedValue;
import android.view.Gravity;
import android.view.MotionEvent;
import android.view.View;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;

import java.util.ArrayList;
import java.util.List;
import java.util.Timer;
import java.util.concurrent.ScheduledThreadPoolExecutor;

import de.devmil.paperlaunch.model.IEntry;
import de.devmil.paperlaunch.utils.BitmapUtils;
import de.devmil.paperlaunch.utils.PositionAndSizeEvaluator;
import de.devmil.paperlaunch.utils.ViewUtils;
import de.devmil.paperlaunch.view.widgets.VerticalTextView;

public class LaunchLaneView extends RelativeLayout {
    interface ILaneListener {
        void onItemSelected(IEntry selectedItem);
        void onStateChanged(LaunchLaneViewModel.State oldState, LaunchLaneViewModel.State newState);
    }

    private LaunchLaneViewModel mViewModel;
    private ILaneListener mLaneListener;

    //view components
    private LinearLayout mSelectIndicatorContainer;
    private LinearLayout mSelectIndicator;
    private LinearLayout mEntriesContainer;
    private ImageView mSelectedIcon;
    private VerticalTextView mSelectedItemTextView;
    private List<LaunchEntryView> mEntryViews = new ArrayList<>();
    private LaunchEntryView mFocusedEntryView;

    public LaunchLaneView(Context context) {
        super(context);
        construct();
    }

    public LaunchLaneView(Context context, AttributeSet attrs) {
        super(context, attrs);
        construct();
    }

    public LaunchLaneView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        construct();
    }

    public LaunchLaneView(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        construct();
    }

    public void doInitializeData(LaunchLaneViewModel viewModel)
    {
        mViewModel = viewModel;

        createViews();
        createEntryViews();
        adaptModelState();
    }

    public void setLaneListener(ILaneListener listener) {
        mLaneListener = listener;
    }

    public void start()
    {
        gotoState(LaunchLaneViewModel.State.Focusing);
    }

    public void stop() {
        removeAllViews();
    }

    public void gotoState(LaunchLaneViewModel.State state)
    {
        transitToState(state);
    }

    public void doHandleTouch(int action, int x, int y)
    {
        int focusSelectionBorder = getWidth();
        if(mViewModel == null) {
            return;
        }
        if(mViewModel.getState() == LaunchLaneViewModel.State.Focusing)
        {
            if(action == MotionEvent.ACTION_UP) {
                sendAllEntriesToState(LaunchEntryViewModel.State.Active, true);
                mFocusedEntryView = null;
            }
            else {
                ensureFocusedEntryAt(y);
                if(mFocusedEntryView != null) {
                    if (mViewModel.isOnRightSide()) {
                        if (x < focusSelectionBorder) {
                            transitToState(LaunchLaneViewModel.State.Selecting);
                        }
                    } else {
                        if (x > focusSelectionBorder)
                            transitToState(LaunchLaneViewModel.State.Selecting);
                    }
                }
            }
        }
        else if(mViewModel.getState() == LaunchLaneViewModel.State.Selected)
        {
            if(mViewModel.isOnRightSide()) {
                if (x > focusSelectionBorder)
                    transitToState(LaunchLaneViewModel.State.Focusing);
            } else {
                if (x < focusSelectionBorder)
                    transitToState(LaunchLaneViewModel.State.Focusing);
            }
        }
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);

        if(mEntryViews != null
                && mEntryViews.size() > 0)
        {
            setMeasuredDimension(mEntryViews.get(0).getMeasuredWidth(), getMeasuredHeight());
        }
    }

    private void construct()
    {
        ViewUtils.disableClipping(this);
    }

    private void createViews()
    {
        removeAllViews();
        mEntriesContainer = new LinearLayout(getContext());
        mEntriesContainer.setOrientation(LinearLayout.VERTICAL);
        RelativeLayout.LayoutParams entriesContainerParams = new LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT);
        entriesContainerParams.addRule(RelativeLayout.ALIGN_PARENT_TOP);
        entriesContainerParams.addRule(RelativeLayout.ALIGN_PARENT_BOTTOM);
        if(mViewModel.isOnRightSide())
        {
            entriesContainerParams.addRule(RelativeLayout.ALIGN_PARENT_RIGHT);
        }
        else
        {
            entriesContainerParams.addRule(RelativeLayout.ALIGN_PARENT_LEFT);
        }

        addView(mEntriesContainer, entriesContainerParams);
        ViewUtils.disableClipping(mEntriesContainer);


        mSelectIndicatorContainer = new LinearLayout(getContext());
        mSelectIndicatorContainer.setOrientation(LinearLayout.VERTICAL);

        RelativeLayout.LayoutParams indicatorContainerParams = new LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT);
        indicatorContainerParams.addRule(RelativeLayout.ALIGN_PARENT_TOP);
        indicatorContainerParams.addRule(RelativeLayout.ALIGN_PARENT_BOTTOM);
        indicatorContainerParams.addRule(RelativeLayout.ALIGN_PARENT_LEFT);
        indicatorContainerParams.addRule(RelativeLayout.ALIGN_PARENT_RIGHT);

        addView(mSelectIndicatorContainer, indicatorContainerParams);
        ViewUtils.disableClipping(mSelectIndicatorContainer);


        mSelectIndicator = new LinearLayout(getContext());
        mSelectIndicator.setBackgroundColor(mViewModel.getFrameDefaultColor());
        mSelectIndicator.setElevation(ViewUtils.getPxFromDip(getContext(), mViewModel.getSelectedImageElevationDip()));
        mSelectIndicator.setVisibility(View.INVISIBLE);
        mSelectIndicator.setGravity(Gravity.CENTER_HORIZONTAL);
        mSelectIndicator.setOrientation(LinearLayout.VERTICAL);

        LinearLayout.LayoutParams selectIndicatorParams = new LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT,
                LinearLayout.LayoutParams.MATCH_PARENT);

        mSelectIndicatorContainer.addView(mSelectIndicator, selectIndicatorParams);
        ViewUtils.disableClipping(mSelectIndicator);

        mSelectedIcon = new ImageView(getContext());
        mSelectedIcon.setScaleType(ImageView.ScaleType.CENTER_INSIDE);
        mSelectedIcon.setImageResource(mViewModel.getUnknownAppImageId());

        LinearLayout.LayoutParams selectIconParams = new LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.WRAP_CONTENT,
                LinearLayout.LayoutParams.WRAP_CONTENT);
        selectIconParams.setMargins(0, (int)ViewUtils.getPxFromDip(getContext(), 5), 0, 0);

        mSelectIndicator.addView(mSelectedIcon, selectIconParams);

        mSelectedItemTextView = new VerticalTextView(getContext());
        mSelectedItemTextView.setVisibility(View.GONE);
        mSelectedItemTextView.setTextSize(TypedValue.COMPLEX_UNIT_SP, mViewModel.getItemNameTextSizeSP());

        LinearLayout.LayoutParams selectedItemTextViewParams = new LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.WRAP_CONTENT,
                LinearLayout.LayoutParams.WRAP_CONTENT);
        selectedItemTextViewParams.setMargins(0, (int)ViewUtils.getPxFromDip(getContext(), 20), 0, 0);
        mSelectIndicator.addView(mSelectedItemTextView, selectedItemTextViewParams);
    }

    private void createEntryViews() {
        mEntriesContainer.removeAllViews();
        mEntryViews.clear();

        for(LaunchEntryViewModel e : mViewModel.getEntries())
        {
            LaunchEntryView ev = new LaunchEntryView(getContext());
            mEntryViews.add(ev);
            LinearLayout.LayoutParams params = new LinearLayout.LayoutParams(
                    LinearLayout.LayoutParams.WRAP_CONTENT,
                    LinearLayout.LayoutParams.WRAP_CONTENT);
            mEntriesContainer.addView(ev, params);

            ev.doInitialize(e);
        }
    }

    private void transitToState(LaunchLaneViewModel.State state)
    {
        switch(state)
        {
            case Init:
                hideSelectionIndicator();
                initEntryState(LaunchEntryViewModel.State.Inactive);
                break;
            case Focusing:
                hideSelectionIndicator();
                sendAllEntriesToState(LaunchEntryViewModel.State.Active, true);
                break;
            case Selecting:
                showSelectionIndicator();
                sendAllEntriesToState(LaunchEntryViewModel.State.Inactive, true, mFocusedEntryView);
                break;
            case Selected:
                fireSelectedEvent();
                break;
        }
        if(mViewModel != null) {
            LaunchLaneViewModel.State oldState = mViewModel.getState();
            LaunchLaneViewModel.State newState = state;
            mViewModel.setState(newState);
            fireStateChangedEvent(oldState, newState);
        }
    }

    private void fireStateChangedEvent(LaunchLaneViewModel.State oldState, LaunchLaneViewModel.State newState) {
        if(mLaneListener != null) {
            mLaneListener.onStateChanged(oldState, newState);
        }
    }

    private void fireSelectedEvent() {
        if(mLaneListener != null) {
            mLaneListener.onItemSelected(mFocusedEntryView.getEntry());
        }
    }

    private void adaptModelState() {
        transitToState(mViewModel.getState());
        applySizeParameters();
    }

    private void initEntryState(LaunchEntryViewModel.State state)
    {
        for(LaunchEntryView ev : mEntryViews)
        {
            ev.setState(state);
        }
    }

    private void sendAllEntriesToState(final LaunchEntryViewModel.State state, boolean topDown)
    {
        sendAllEntriesToState(state, topDown, null);
    }

    private void sendAllEntriesToState(final LaunchEntryViewModel.State state, boolean topDown, LaunchEntryView except)
    {
        int delay = 0;
        int start = 0;
        int end = mEntryViews.size();
        int diff = 1;
        if(!topDown)
        {
            start = mEntryViews.size() - 1;
            end = -1;
            diff = -1;
        }
        for(int i=start; i<end; i+=diff)
        {
            if(mEntryViews.get(i) != except) {
                mEntryViews.get(i).gotoState(state, delay += mViewModel.getEntryMoveDiffMS());
            }
        }
    }

    private void applySizeParameters()
    {
        mSelectedIcon.setMaxHeight((int) ViewUtils.getPxFromDip(getContext(), mViewModel.getImageWidthDip()));
        mSelectedIcon.setMaxWidth((int) ViewUtils.getPxFromDip(getContext(), mViewModel.getImageWidthDip()));
    }

    private void showSelectionIndicator()
    {
        if(mFocusedEntryView == null) {
            return;
        }
        Rect fromRect = new Rect();
        mFocusedEntryView.getHitRect(fromRect);
        Rect toRect = new Rect();
        mSelectIndicatorContainer.getHitRect(toRect);

        Drawable drawable = mFocusedEntryView.getEntry().getIcon(getContext());
        mSelectedIcon.setImageDrawable(drawable);

        mSelectedItemTextView.setText(mFocusedEntryView.getEntry().getName(getContext()));

        Palette p = Palette.from(BitmapUtils.drawableToBitmap(drawable)).generate();

        mSelectIndicator.setBackgroundColor(getColorFromPalette(p, mViewModel.getFrameDefaultColor()));

        try {
            ObjectAnimator anim = ObjectAnimator.ofObject(
                    mSelectIndicator,
                    "margins",
                    new PositionAndSizeEvaluator(mSelectIndicator),
                    fromRect,
                    toRect);
            anim.addListener(new Animator.AnimatorListener() {
                @Override
                public void onAnimationStart(Animator animation) {
                }

                @Override
                public void onAnimationEnd(Animator animation) {
                    transitToState(LaunchLaneViewModel.State.Selected);
                }

                @Override
                public void onAnimationCancel(Animator animation) {
                }

                @Override
                public void onAnimationRepeat(Animator animation) {
                }
            });
            anim.setDuration(mViewModel.getSelectingAnimationDurationMS());
            anim.start();

            new Thread(new Runnable() {
                @Override
                public void run() {
                    try {
                        Thread.sleep(mViewModel.getSelectingAnimationDurationMS() / 2);
                        mSelectedItemTextView.post(new Runnable() {
                            @Override
                            public void run() {
                                mSelectedItemTextView.setVisibility(View.VISIBLE);
                            }
                        });
                    } catch (InterruptedException e) {
                    }
                }
            }).start();
        } catch(Exception e) {
        }
        mSelectIndicator.setVisibility(View.VISIBLE);
    }

    private int getColorFromPalette(Palette palette, int defaultColor) {
        int result = palette.getLightMutedColor(Color.BLACK);
        if(result == Color.BLACK) {
            result = palette.getLightVibrantColor(Color.BLACK);
        }
        if(result == Color.BLACK) {
            result = palette.getMutedColor(Color.BLACK);
        }
        if(result == Color.BLACK) {
            result = palette.getVibrantColor(Color.BLACK);
        }
        if(result == Color.BLACK) {
            result = defaultColor;
        }
        return result;
    }

    private void hideSelectionIndicator()
    {
        if(mSelectIndicator != null) {
            mSelectIndicator.setVisibility(View.INVISIBLE);
        }
        if(mSelectedItemTextView != null) {
            mSelectedItemTextView.setVisibility(View.GONE);
        }
    }

    private void ensureFocusedEntryAt(int y)
    {
        mFocusedEntryView = null;
        for(LaunchEntryView ev : mEntryViews)
        {
            boolean hit = isEntryAt(ev, y);
            LaunchEntryViewModel.State desiredState = LaunchEntryViewModel.State.Active;
            if(hit)
            {
                desiredState = LaunchEntryViewModel.State.Focused;
                mFocusedEntryView = ev;
            }
            ev.gotoState(desiredState);
        }
    }

    private boolean isEntryAt(LaunchEntryView entryView, int y)
    {
        return entryView.getY() < y && y < entryView.getY() + entryView.getHeight();
    }
}

File: app/src/main/java/de/devmil/paperlaunch/view/LaunchLaneView.kt
/*
 * Copyright 2015 Devmil Solutions
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package de.devmil.paperlaunch.view

import android.animation.Animator
import android.animation.ObjectAnimator
import android.content.Context
import android.graphics.Rect
import android.support.v4.content.ContextCompat
import android.util.AttributeSet
import android.util.TypedValue
import android.view.Gravity
import android.view.MotionEvent
import android.view.View
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.RelativeLayout

import java.util.ArrayList

import de.devmil.paperlaunch.R
import de.devmil.paperlaunch.config.LauncherGravity
import de.devmil.paperlaunch.model.IEntry
import de.devmil.paperlaunch.utils.BitmapUtils
import de.devmil.paperlaunch.utils.PositionAndSizeEvaluator
import de.devmil.paperlaunch.view.utils.ViewUtils
import de.devmil.paperlaunch.view.utils.ColorUtils
import de.devmil.paperlaunch.view.widgets.VerticalTextView

class LaunchLaneView : RelativeLayout {
    interface ILaneListener {
        fun onItemSelected(selectedItem: IEntry?)
        fun onItemSelecting(selectedItem: IEntry?)
        fun onStateChanged(oldState: LaunchLaneViewModel.State, newState: LaunchLaneViewModel.State)
    }

    private var viewModel: LaunchLaneViewModel? = null
    private var laneListener: ILaneListener? = null

    //view components
    private var selectIndicatorContainer: LinearLayout? = null
    private var selectIndicator: LinearLayout? = null
    private var entriesContainer: LinearLayout? = null
    private var selectedIcon: ImageView? = null
    private var selectedItemTextView: VerticalTextView? = null
    private val entryViews = ArrayList<LaunchEntryView>()
    private var focusedEntryView: LaunchEntryView? = null

    constructor(context: Context) : super(context) {
        construct()
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        construct()
    }

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        construct()
    }

    @Suppress("unused")
    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int, defStyleRes: Int) : super(context, attrs, defStyleAttr, defStyleRes) {
        construct()
    }

    fun doInitializeData(viewModel: LaunchLaneViewModel) {
        this.viewModel = viewModel

        createViews()
        createEntryViews()
        adaptModelState()
    }

    fun setLaneListener(listener: ILaneListener) {
        laneListener = listener
    }

    fun start() {
        gotoState(LaunchLaneViewModel.State.Focusing)
    }

    fun stop() {
        removeAllViews()
        viewModel?.let {
            it.state = LaunchLaneViewModel.State.Init
        }
    }

    private fun gotoState(state: LaunchLaneViewModel.State) {
        transitToState(state)
    }

    fun doHandleTouch(action: Int, x: Int, y: Int) {
        val focusSelectionBorder = width
        viewModel?.let {
            if (it.state === LaunchLaneViewModel.State.Focusing) {
                if (action == MotionEvent.ACTION_UP) {
                    sendAllEntriesToState(LaunchEntryViewModel.State.Active)
                    focusedEntryView = null
                } else {
                    ensureFocusedEntryAt(y)
                    if (focusedEntryView != null) {
                        if (it.isOnRightSide) {
                            if (x < focusSelectionBorder) {
                                transitToState(LaunchLaneViewModel.State.Selecting)
                            }
                        } else {
                            if (x > 0) {
                                transitToState(LaunchLaneViewModel.State.Selecting)
                            }
                        }
                    }
                }
            } else if (it.state === LaunchLaneViewModel.State.Selected) {
                if (it.isOnRightSide) {
                    if (x > focusSelectionBorder)
                        transitToState(LaunchLaneViewModel.State.Focusing)
                } else {
                    if (x < 0)
                        transitToState(LaunchLaneViewModel.State.Focusing)
                }
            }
        }
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)

        if (entryViews.size > 0) {
            setMeasuredDimension(entryViews[0].measuredWidth, measuredHeight)
        }
    }

    private fun construct() {
        ViewUtils.disableClipping(this)
    }

    private fun createViews() {
        removeAllViews()
        val localEntriesContainer = LinearLayout(context)
        localEntriesContainer.orientation = LinearLayout.VERTICAL

        val localViewModel = viewModel!!

        when (localViewModel.launcherGravity) {
            LauncherGravity.Top -> localEntriesContainer.gravity = Gravity.TOP
            LauncherGravity.Center -> localEntriesContainer.gravity = Gravity.CENTER_VERTICAL
            LauncherGravity.Bottom -> localEntriesContainer.gravity = Gravity.BOTTOM
        }

        val entriesContainerParams = RelativeLayout.LayoutParams(
                RelativeLayout.LayoutParams.WRAP_CONTENT,
                RelativeLayout.LayoutParams.WRAP_CONTENT)
        entriesContainerParams.addRule(RelativeLayout.ALIGN_PARENT_TOP)
        entriesContainerParams.addRule(RelativeLayout.ALIGN_PARENT_BOTTOM)
        if (localViewModel.isOnRightSide) {
            entriesContainerParams.addRule(RelativeLayout.ALIGN_PARENT_RIGHT)
        } else {
            entriesContainerParams.addRule(RelativeLayout.ALIGN_PARENT_LEFT)
        }

        addView(localEntriesContainer, entriesContainerParams)
        ViewUtils.disableClipping(localEntriesContainer)


        val localSelectIndicatorContainer = LinearLayout(context)
        localSelectIndicatorContainer.orientation = LinearLayout.VERTICAL

        val indicatorContainerParams = RelativeLayout.LayoutParams(
                RelativeLayout.LayoutParams.WRAP_CONTENT,
                RelativeLayout.LayoutParams.WRAP_CONTENT)
        indicatorContainerParams.addRule(RelativeLayout.ALIGN_PARENT_TOP)
        indicatorContainerParams.addRule(RelativeLayout.ALIGN_PARENT_BOTTOM)
        indicatorContainerParams.addRule(RelativeLayout.ALIGN_PARENT_LEFT)
        indicatorContainerParams.addRule(RelativeLayout.ALIGN_PARENT_RIGHT)

        addView(localSelectIndicatorContainer, indicatorContainerParams)
        ViewUtils.disableClipping(localSelectIndicatorContainer)


        val localSelectIndicator = LinearLayout(context)
        localSelectIndicator.setBackgroundColor(localViewModel.frameDefaultColor)
        localSelectIndicator.elevation = ViewUtils.getPxFromDip(context, localViewModel.selectedImageElevationDip)
        localSelectIndicator.visibility = View.INVISIBLE
        localSelectIndicator.gravity = Gravity.CENTER_HORIZONTAL
        localSelectIndicator.orientation = LinearLayout.VERTICAL

        val selectIndicatorParams = LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT,
                LinearLayout.LayoutParams.MATCH_PARENT)

        localSelectIndicatorContainer.addView(localSelectIndicator, selectIndicatorParams)
        ViewUtils.disableClipping(localSelectIndicator)

        val localSelectedIcon = ImageView(context)
        localSelectedIcon.scaleType = ImageView.ScaleType.CENTER_INSIDE
        localSelectedIcon.setImageResource(localViewModel.unknownAppImageId)

        val selectIconParams = LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.WRAP_CONTENT,
                LinearLayout.LayoutParams.WRAP_CONTENT)
        selectIconParams.setMargins(0, ViewUtils.getPxFromDip(context, localViewModel.laneIconTopMarginDip).toInt(), 0, 0)

        localSelectIndicator.addView(localSelectedIcon, selectIconParams)

        val localSelectedItemTextView = VerticalTextView(context)
        localSelectedItemTextView.visibility = View.GONE
        localSelectedItemTextView.setTextSize(TypedValue.COMPLEX_UNIT_SP, localViewModel.itemNameTextSizeSP)
        //this is needed because the parts in the system run with another theme than the application parts
        localSelectedItemTextView.setTextColor(ContextCompat.getColor(context, R.color.name_label))

        val selectedItemTextViewParams = LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.WRAP_CONTENT,
                LinearLayout.LayoutParams.WRAP_CONTENT)
        selectedItemTextViewParams.setMargins(0, ViewUtils.getPxFromDip(context, localViewModel.laneTextTopMarginDip).toInt(), 0, 0)
        localSelectIndicator.addView(localSelectedItemTextView, selectedItemTextViewParams)

        entriesContainer = localEntriesContainer
        selectIndicatorContainer = localSelectIndicatorContainer
        selectIndicator = localSelectIndicator
        selectedIcon = localSelectedIcon
        selectedItemTextView = localSelectedItemTextView
    }

    private fun createEntryViews() {
        val localEntriesContainer = entriesContainer!!
        val localViewModel = viewModel!!
        localEntriesContainer.removeAllViews()
        entryViews.clear()

        val entries = localViewModel.entries
        for (e in entries) {
            val ev = LaunchEntryView(context)
            entryViews.add(ev)
            val params = LinearLayout.LayoutParams(
                    LinearLayout.LayoutParams.WRAP_CONTENT,
                    LinearLayout.LayoutParams.WRAP_CONTENT)
            localEntriesContainer.addView(ev, params)

            ev.doInitialize(e)
        }
    }

    private fun transitToState(state: LaunchLaneViewModel.State) {
        when (state) {
            LaunchLaneViewModel.State.Init -> {
                fireNotSelectedEvents()
                hideSelectionIndicator()
                hideEntries()
                initEntryState(LaunchEntryViewModel.State.Inactive)
            }
            LaunchLaneViewModel.State.Focusing -> {
                fireNotSelectedEvents()
                hideSelectionIndicator()
                showEntries()
                sendAllEntriesToState(LaunchEntryViewModel.State.Active)
            }
            LaunchLaneViewModel.State.Selecting -> {
                showSelectionIndicator()
                sendAllEntriesToState(LaunchEntryViewModel.State.Inactive, focusedEntryView)
                fireSelectingEvent()
            }
            LaunchLaneViewModel.State.Selected -> fireSelectedEvent()
        }
        viewModel?.let {
            val oldState = it.state
            it.state = state
            fireStateChangedEvent(oldState, state)
        }
    }

    private fun fireStateChangedEvent(oldState: LaunchLaneViewModel.State, newState: LaunchLaneViewModel.State) {
        laneListener?.onStateChanged(oldState, newState)
    }

    private fun fireSelectedEvent() {
        laneListener?.let { listener ->
            focusedEntryView?.let { entryView ->
                listener.onItemSelected(entryView.entry)
            }
        }
    }

    private fun fireSelectingEvent() {
        laneListener?.let { listener ->
            focusedEntryView?.let { entryView ->
                listener.onItemSelecting(entryView.entry)
            }
        }
    }

    private fun fireNotSelectedEvents() {
        laneListener?.let { listener ->
            listener.onItemSelected(null)
            listener.onItemSelecting(null)
        }
    }

    private fun adaptModelState() {
        transitToState(viewModel!!.state)
        applySizeParameters()
    }

    private fun initEntryState(state: LaunchEntryViewModel.State) {
        for (ev in entryViews) {
            ev.setState(state)
        }
    }

    private fun sendAllEntriesToState(state: LaunchEntryViewModel.State, except: LaunchEntryView? = null) {
        val localViewModel = viewModel!!
        var delay = 0

        val entryCount = entryViews.size
        val count = entryCount / 2
        var centerIndex = -1
        if (entryViews.size % 2 != 0) {
            centerIndex = count
        }

        if (centerIndex >= 0) {
            if (entryViews[centerIndex] != except) {
                entryViews[centerIndex].gotoState(state, delay)
            }
            delay += localViewModel.entryMoveDiffMS
        }

        for (i in count - 1 downTo 0) {
            @Suppress("UnnecessaryVariable")
            val upperIdx = i
            val lowerIdx = entryCount - 1 - i
            if (entryViews[upperIdx] != except) {
                entryViews[upperIdx].gotoState(state, delay)
            }
            if (entryViews[lowerIdx] != except) {
                entryViews[lowerIdx].gotoState(state, delay)
            }

            delay += localViewModel.entryMoveDiffMS
        }
    }

    private fun applySizeParameters() {
        val localViewModel = viewModel!!
        val localSelectedIcon = selectedIcon!!
        localSelectedIcon.maxHeight = ViewUtils.getPxFromDip(context, localViewModel.imageWidthDip).toInt()
        localSelectedIcon.maxWidth = ViewUtils.getPxFromDip(context, localViewModel.imageWidthDip).toInt()
    }

    private fun showSelectionIndicator() {
        if (focusedEntryView == null) {
            return
        }
        val localFocusedEntryView = focusedEntryView!!
        val localSelectIndicatorContainer = selectIndicatorContainer!!
        val localSelectedIcon = selectedIcon!!
        val localSelectedItemTextView = selectedItemTextView!!
        val localSelectIndicator = selectIndicator!!
        val localViewModel = viewModel!!

        val fromRect = Rect()
        localFocusedEntryView.getHitRect(fromRect)
        val toRect = Rect()
        localSelectIndicatorContainer.getHitRect(toRect)

        val drawable = localFocusedEntryView.entry.getIcon(context)
        localSelectedIcon.setImageDrawable(drawable)

        val useIconColor = localFocusedEntryView.entry.useIconColor()

        localSelectedItemTextView.text = localFocusedEntryView.entry.getName(context)

        val bmpResult = BitmapUtils.drawableToBitmap(drawable)
        if (useIconColor
                && bmpResult != null) {
            localSelectIndicator.setBackgroundColor(
                    ColorUtils.getBackgroundColorFromImage(
                            bmpResult.bitmap,
                            localViewModel.frameDefaultColor))
        } else {
            localSelectIndicator.setBackgroundColor(
                    localViewModel.frameDefaultColor)
        }

        if (bmpResult != null && bmpResult.isNew) {
            bmpResult.bitmap.recycle()
        }

        try {
            val anim = ObjectAnimator.ofObject(
                    localSelectIndicator,
                    "margins",
                    PositionAndSizeEvaluator(localSelectIndicator),
                    fromRect,
                    toRect)
            anim.addListener(object : Animator.AnimatorListener {
                override fun onAnimationStart(animation: Animator) {}

                override fun onAnimationEnd(animation: Animator) {
                    transitToState(LaunchLaneViewModel.State.Selected)
                }

                override fun onAnimationCancel(animation: Animator) {}

                override fun onAnimationRepeat(animation: Animator) {}
            })
            anim.duration = localViewModel.selectingAnimationDurationMS.toLong()
            anim.start()

            Thread(Runnable {
                try {
                    Thread.sleep((localViewModel.selectingAnimationDurationMS / 2).toLong())
                    localSelectedItemTextView.post { localSelectedItemTextView.visibility = View.VISIBLE }
                } catch (e: InterruptedException) {
                }
            }).start()
        } catch (e: Exception) {
        }

        localSelectIndicator.visibility = View.VISIBLE
    }

    private fun hideSelectionIndicator() {
        selectIndicator?.visibility = View.INVISIBLE
        selectedItemTextView?.visibility = View.GONE
    }

    private fun hideEntries() {
        for (ev in entryViews) {
            ev.visibility = View.INVISIBLE
        }
    }

    private fun showEntries() {
        for (ev in entryViews) {
            ev.visibility = View.VISIBLE
        }
    }

    private fun ensureFocusedEntryAt(y: Int) {
        focusedEntryView = null
        for (ev in entryViews) {
            val hit = isEntryAt(ev, y)
            var desiredState: LaunchEntryViewModel.State = LaunchEntryViewModel.State.Active
            if (hit) {
                desiredState = LaunchEntryViewModel.State.Focused
                focusedEntryView = ev
            }
            ev.gotoState(desiredState)
        }
    }

    private fun isEntryAt(entryView: LaunchEntryView, y: Int): Boolean {
        return entryView.y < y && y < entryView.y + entryView.height
    }
}
